File: flut/flutter/widgets/drag_target.py
from typing import Callable, Optional, override

from flut._flut.engine import FlutValueObject, _flut_pack_value
from flut._flut.runtime import (
    _flut_unpack_required_field,
    _flut_unpack_optional_field,
)
from flut.dart.ui import Offset
from flut.flutter.gestures.velocity_tracker import Velocity
from flut.flutter.painting.basic_types import Axis
from flut.flutter.rendering.proxy_box import HitTestBehavior
from flut.flutter.widgets.framework import Widget, BuildContext


class DraggableDetails(FlutValueObject):
    _flut_type = "DraggableDetails"

    def __init__(
        self,
        *,
        wasAccepted: bool = False,
        velocity: Velocity,
        offset: Offset,
    ):
        super().__init__()
        self.wasAccepted = wasAccepted
        self.velocity = velocity
        self.offset = offset

    @staticmethod
    def _flut_unpack(data: dict) -> "DraggableDetails":
        return DraggableDetails(
            wasAccepted=_flut_unpack_required_field(data, "wasAccepted"),
            velocity=_flut_unpack_required_field(data, "velocity"),
            offset=_flut_unpack_required_field(data, "offset"),
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["wasAccepted"] = _flut_pack_value(self.wasAccepted)
        result["velocity"] = _flut_pack_value(self.velocity)
        result["offset"] = _flut_pack_value(self.offset)
        return result


class DragTargetDetails(FlutValueObject):
    _flut_type = "DragTargetDetails"

    def __init__(
        self,
        *,
        data,
        offset: Offset,
    ):
        super().__init__()
        self.data = data
        self.offset = offset

    @staticmethod
    def _flut_unpack(data: dict) -> "DragTargetDetails":
        return DragTargetDetails(
            data=_flut_unpack_required_field(data, "data"),
            offset=_flut_unpack_required_field(data, "offset"),
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["data"] = _flut_pack_value(self.data)
        result["offset"] = _flut_pack_value(self.offset)
        return result


def wrap_drag_target_builder(builder):
    def _wrapper(context, candidate_data, rejected_data):
        built = builder(context, candidate_data, rejected_data)
        if built is not None:
            return built._flut_pack()
        return None

    return _wrapper


class Draggable(Widget):
    _flut_type = "Draggable"

    def __init__(
        self,
        *,
        key=None,
        data=None,
        axis: Optional[Axis] = None,
        childWhenDragging: Optional[Widget] = None,
        feedbackOffset: Offset = Offset(),
        affinity: Optional[Axis] = None,
        maxSimultaneousDrags: Optional[int] = None,
        onDragStarted: Optional[Callable[[], None]] = None,
        onDragUpdate=None,
        onDraggableCanceled=None,
        onDragEnd=None,
        onDragCompleted: Optional[Callable[[], None]] = None,
        ignoringFeedbackSemantics: bool = True,
        ignoringFeedbackPointer: bool = True,
        rootOverlay: bool = False,
        hitTestBehavior: HitTestBehavior = HitTestBehavior.deferToChild,
        child: Widget,
        feedback: Widget,
    ):
        super().__init__(key=key)
        self.data = data
        self.axis = axis
        self.childWhenDragging = childWhenDragging
        self.feedbackOffset = feedbackOffset
        self.affinity = affinity
        self.maxSimultaneousDrags = maxSimultaneousDrags
        self.onDragStarted = onDragStarted
        self.onDragUpdate = onDragUpdate
        self.onDraggableCanceled = onDraggableCanceled
        self.onDragEnd = onDragEnd
        self.onDragCompleted = onDragCompleted
        self.ignoringFeedbackSemantics = ignoringFeedbackSemantics
        self.ignoringFeedbackPointer = ignoringFeedbackPointer
        self.rootOverlay = rootOverlay
        self.hitTestBehavior = hitTestBehavior
        self.child = child
        self.feedback = feedback

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["feedbackOffset"] = _flut_pack_value(self.feedbackOffset)
        result["ignoringFeedbackSemantics"] = _flut_pack_value(
            self.ignoringFeedbackSemantics
        )
        result["ignoringFeedbackPointer"] = _flut_pack_value(
            self.ignoringFeedbackPointer
        )
        result["rootOverlay"] = _flut_pack_value(self.rootOverlay)
        result["hitTestBehavior"] = _flut_pack_value(self.hitTestBehavior)
        result["child"] = _flut_pack_value(self.child)
        result["feedback"] = _flut_pack_value(self.feedback)
        if self.data is not None:
            result["data"] = _flut_pack_value(self.data)
        if self.axis is not None:
            result["axis"] = _flut_pack_value(self.axis)
        if self.childWhenDragging is not None:
            result["childWhenDragging"] = _flut_pack_value(self.childWhenDragging)
        if self.affinity is not None:
            result["affinity"] = _flut_pack_value(self.affinity)
        if self.maxSimultaneousDrags is not None:
            result["maxSimultaneousDrags"] = _flut_pack_value(self.maxSimultaneousDrags)
        if self.onDragStarted is not None:
            result["onDragStarted"] = self._register_action(
                self.onDragStarted, "VoidCallback"
            )._flut_pack()
        if self.onDragUpdate is not None:
            result["onDragUpdate"] = self._register_action(
                self.onDragUpdate, "GestureDragUpdateCallback"
            )._flut_pack()
        if self.onDraggableCanceled is not None:
            result["onDraggableCanceled"] = self._register_action(
                self.onDraggableCanceled, "DraggableCanceledCallback"
            )._flut_pack()
        if self.onDragEnd is not None:
            result["onDragEnd"] = self._register_action(
                self.onDragEnd, "DragEndCallback"
            )._flut_pack()
        if self.onDragCompleted is not None:
            result["onDragCompleted"] = self._register_action(
                self.onDragCompleted, "VoidCallback"
            )._flut_pack()
        return result


class DragTarget(Widget):
    _flut_type = "DragTarget"

    def __init__(
        self,
        *,
        key=None,
        builder: Callable[[BuildContext, list, list], Widget],
        onWillAcceptWithDetails=None,
        onAcceptWithDetails=None,
        onLeave=None,
        onMove=None,
        hitTestBehavior: HitTestBehavior = HitTestBehavior.translucent,
    ):
        super().__init__(key=key)
        self.builder = builder
        self.onWillAcceptWithDetails = onWillAcceptWithDetails
        self.onAcceptWithDetails = onAcceptWithDetails
        self.onLeave = onLeave
        self.onMove = onMove
        self.hitTestBehavior = hitTestBehavior

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["builder"] = self._register_build_action(
            wrap_drag_target_builder(self.builder), "DragTargetBuilder"
        )._flut_pack()
        result["hitTestBehavior"] = _flut_pack_value(self.hitTestBehavior)
        if self.onWillAcceptWithDetails is not None:
            result["onWillAcceptWithDetails"] = self._register_action(
                self.onWillAcceptWithDetails, "DragTargetWillAcceptWithDetails"
            )._flut_pack()
        if self.onAcceptWithDetails is not None:
            result["onAcceptWithDetails"] = self._register_action(
                self.onAcceptWithDetails, "DragTargetAcceptWithDetails"
            )._flut_pack()
        if self.onLeave is not None:
            result["onLeave"] = self._register_action(
                self.onLeave, "DragTargetLeave"
            )._flut_pack()
        if self.onMove is not None:
            result["onMove"] = self._register_action(
                self.onMove, "DragTargetMove"
            )._flut_pack()
        return result

File: flut/.flutter/lib/flutter/widgets/drag_target.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/dart/ui.dart';
import 'package:flut/flutter/gestures/velocity_tracker.dart';

class FlutDraggableDetails extends FlutValueObject {
  final DraggableDetails details;

  const FlutDraggableDetails(this.details) : super('DraggableDetails');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    result['wasAccepted'] = details.wasAccepted;
    result['velocity'] = FlutVelocity(details.velocity).flutEncode();
    result['offset'] = FlutOffset(details.offset).flutEncode();
    return result;
  }

  static DraggableDetails? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return DraggableDetails(
      wasAccepted: runtime.unpackRequiredField<bool>(data, 'wasAccepted'),
      velocity: runtime.unpackRequiredField<Velocity>(data, 'velocity'),
      offset: runtime.unpackRequiredField<Offset>(data, 'offset'),
    );
  }
}

class FlutDragTargetDetails extends FlutValueObject {
  final DragTargetDetails details;
  final FlutRuntime? _runtime;

  const FlutDragTargetDetails(this.details, [this._runtime])
    : super('DragTargetDetails');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (_runtime != null) {
      result['data'] = _runtime.encodeValue(details.data);
    }
    result['offset'] = FlutOffset(details.offset).flutEncode();
    return result;
  }

  static DragTargetDetails? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return DragTargetDetails(
      data: runtime.unpackDynamicRequiredField(data, 'data'),
      offset: runtime.unpackRequiredField<Offset>(data, 'offset'),
    );
  }
}

class FlutDraggable {
  FlutDraggable._();

  static Draggable<Object>? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return Draggable<Object>(
      key: runtime.decodeKey(data),
      data: runtime.unpackDynamicOptionalField(data, 'data'),
      axis: runtime.unpackOptionalField<Axis>(data, 'axis'),
      childWhenDragging: runtime.unpackOptionalField<Widget>(
        data,
        'childWhenDragging',
      ),
      feedbackOffset: runtime.unpackRequiredField<Offset>(
        data,
        'feedbackOffset',
      ),
      affinity: runtime.unpackOptionalField<Axis>(data, 'affinity'),
      maxSimultaneousDrags: runtime.unpackOptionalField<int>(
        data,
        'maxSimultaneousDrags',
      ),
      onDragStarted: runtime.unpackOptionalCallback(data, 'onDragStarted'),
      onDragUpdate: runtime.unpackOptionalCallback(data, 'onDragUpdate'),
      onDraggableCanceled: runtime.unpackOptionalCallback(
        data,
        'onDraggableCanceled',
      ),
      onDragEnd: runtime.unpackOptionalCallback(data, 'onDragEnd'),
      onDragCompleted: runtime.unpackOptionalCallback(data, 'onDragCompleted'),
      ignoringFeedbackSemantics: runtime.unpackRequiredField<bool>(
        data,
        'ignoringFeedbackSemantics',
      ),
      ignoringFeedbackPointer: runtime.unpackRequiredField<bool>(
        data,
        'ignoringFeedbackPointer',
      ),
      rootOverlay: runtime.unpackRequiredField<bool>(data, 'rootOverlay'),
      hitTestBehavior: runtime.unpackRequiredField<HitTestBehavior>(
        data,
        'hitTestBehavior',
      ),
      feedback: runtime.unpackRequiredField<Widget>(data, 'feedback'),
      child: runtime.unpackRequiredField<Widget>(data, 'child'),
    );
  }
}

class FlutDragTarget {
  FlutDragTarget._();

  static DragTarget<Object>? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return DragTarget<Object>(
      key: runtime.decodeKey(data),
      builder: runtime.unpackRequiredCallback(data, 'builder'),
      onWillAcceptWithDetails: runtime.unpackOptionalCallback(
        data,
        'onWillAcceptWithDetails',
      ),
      onAcceptWithDetails: runtime.unpackOptionalCallback(
        data,
        'onAcceptWithDetails',
      ),
      onLeave: runtime.unpackOptionalCallback(data, 'onLeave'),
      onMove: runtime.unpackOptionalCallback(data, 'onMove'),
      hitTestBehavior: runtime.unpackRequiredField<HitTestBehavior>(
        data,
        'hitTestBehavior',
      ),
    );
  }
}
